allow to edit an existing resource

adietish opened the pull request:
fixes #175

adietish commented:
When editing resources in intellij-tekton one has confirm a dialog which warns of editing a non-project. This happens once per IJ session for each resource. When using temp-files like in intellij-tekton I faced the same annoyance. See here:

https://user-images.githubusercontent.com/25126/112275902-6568c680-8c80-11eb-90d8-1eb5a3b77804.mov

These dialogs started to appear (especially on MacOS) when editing resources started to use temporary files instead of scratches: https://github.com/redhat-developer/intellij-tekton/pull/111/commits/1fdff380838614bf01dafade754dec1ba6fedbae#diff-ee479dbd6ef04cbfce979e80e04c42b1c120339a815475be45b2569228565760L71
The reasoning for this changes was that the scratches could not be edited with the backing of a schema.
I found out that scratches were created using ```Language.ANY```.
```
VirtualFile vf = ScratchRootType.getInstance().createScratchFile(project, name, Language.ANY, content)
```
Using ```YAMLLanguage.INSTANCE``` instead has my scratches come up with qualified YAML support when editing:
```
ScratchRootType.getInstance().createScratchFile(project, name, YAMLLanguage.INSTANCE, Serialization.asYaml(resource))
```
Using scratches does not cause the annoying prompts to confirm editing non-projects files to appear. I therefore used scratches in this attempt

adietish commented:
Schema validation seems not to work as pointed out in the [PR for tekton](https://github.com/redhat-developer/intellij-tekton/commit/1fdff380838614bf01dafade754dec1ba6fedbae#diff-ee479dbd6ef04cbfce979e80e04c42b1c120339a815475be45b2569228565760L71). Investigating.

adietish commented:
found out why schema validation doesn’t happen:
```
public class YamlJsonEnabler implements JsonSchemaEnabler {
  @Override
  public boolean isEnabledForFile(VirtualFile file) {
    FileType type = file.getFileType();
    return type instanceof LanguageFileType && ((LanguageFileType)type).getLanguage() instanceof YAMLLanguage;
  }
```
Yaml schema validator checks the type of the file to say whether he’s enabled or not for a given file, it'll answer `true` for for a `FileType` who's language is an instance of  `YAMLLanguage`.
The file type of a “real” yaml file (ex. /tmp/test.yml) is `YAMLFileType`. A scratch file is `ScratchFileType` who's language type always is `PlainTextLanguage` even if it’s content is YAML.

adietish commented:
This shortcoming is a known issue for IJ: https://youtrack.jetbrains.com/issue/IDEA-233082. I added my findings in their jira.
There's a workaround suggested in https://youtrack.jetbrains.com/issue/IDEA-258170 which is not possible for us because it causes code completion etc. not to work.
Overriding the `YamlJsonEnabler` doesn't work either:
> unfortunately YamlJsonEnabler has no set id, otherwise it could work by adding custom one and setting before <id>, unfortunately registering with order=first is not guaranteed to work 100%
Overriding the file type for our yaml scratch files would potentially break other things:
> you cannot really override an existing filetype mapping, all sorts of things will break
So we're left using real files in /tmp as we did so far. To avoid the "edit non-project files" dialog, jetbrains devs suggested an extension point:
> `com.intellij.openapi.fileEditor.impl.NonProjectFileWritingAccessExtension` might help for temp file

jeffmaury commented:
I'm kinda lost here: it is happening only on MacOS because I don't remember we had it on IJ Tekton or is it related to IJ version (as your have a minimum of 2018.3 and we have 2019.3). Worth trying on a more recent version

adietish commented:
> I'm kinda lost here: it is happening only on MacOS because I don't remember we had it on IJ Tekton or is it related to IJ version (as your have a minimum of 2018.3 and we have 2019.3). Worth trying on a more recent version

Me and Yevhen on Mac, we have this dialog all the time. @lstocchi reports to have it here and there. On Linux (and maybe on Windows?) this never happens if the temp directory is at /tmp, while it happens when it is at /var/tmp (?).

@lstocchi you might enlighten @jeffmaury on when you saw the dialog?

lstocchi commented:
I never really faced that issue with our current code in IJ Tekton. I was able to replicate it while i was trying other things to get rid of that. For example, by saving a file inside the idea-sandbox folder (`PathManager.getPluginTempPath`) i had that dialog.

adietish commented:
@jeffmaury, @lstocchi this can be tested. Looks to me as if it works. Intensive testing needed for the non-happy paths though.
The editor is watching the resource on cluster so that it can detect removal & modifications. It would then in such cases display a notification:
![image](https://user-images.githubusercontent.com/25126/116404085-95325d80-a82e-11eb-98ce-87507ffac213.png)

adietish commented:
@jeffmaury, @lstocchi: just found a bug: switching context/namespace causes editor to error when saving.

adietish commented:
@jeffmaury, @lstocchi: fixed the error when switching namespace, context while keeping editor opened. There's still an issue/limitation remaining:
Open editors are restored when you launch IJ. We have no idea what context was used for them. If the context (as in kubeconfig) is still the same, we're fine, saving works. If it isn't we're screwed, saving wont work. To overcome this we would have to store the context (cluster, user, namespace) that was used when this editor was (manually) opened. Things could still go wrong in that case if the context was deleted from kubeconfig. Overcoming this would require us to store the whole kubeconfig with the editor so that it can be used when the editor is restored.

adietish commented:
Still other errors present. Continuing fixing stuff.

lstocchi commented:
> @jeffmaury, @lstocchi: fixed the error when switching namespace, context while keeping editor opened. There's still an issue/limitation remaining though:
> Open editors are restored when you launch IJ. We have no idea what context was used when they were manually opened. If the current context (as in kubeconfig) is still the same, we're fine, saving works. If it was changed, we're screwed, saving wont work - chaning the current namespace doesnt matter. To overcome this we would have to store the context (cluster, user, namespace) that was used when this editor was (manually) opened. Things could still go wrong in that case if the context was deleted from kubeconfig. Overcoming this would require us to store a copy of the kubeconfig (as the context entity is referencing other entities like user etc.) with the editor so that it can be used when the editor is restored.
> Thoughts?

Luckily for me i have a strong heart but looking at the files changed really made me almost passed out :rofl: 
BTW, the other day i was looking for something similar and i found https://intellij-support.jetbrains.com/hc/en-us/community/posts/206824755-Persist-User-Data-on-a-Virtual-Fil, then i didn't try it but maybe it's worth a try. Not sure if you already saw this.

adietish commented:
@lstocchi 
> Luckily for me i have a strong heart but looking at the files changed really made me almost passed out 🤣
LOL, what makes this change huge are the tons of schemas that I added, 400k lines 🤣 

> BTW, the other day i was looking for something similar and i found https://intellij-support.jetbrains.com/hc/en-us/community/posts/206824755-Persist-User-Data-on-a-Virtual-Fil, then i didn't try it but maybe it's worth a try. Not sure if you already saw this.

excellent, thx for the hint! looks exactly like what I could use. Still not sure if we should do store the kubeconfig with the editor though. Maybe that's something for a latter improvement, limiting saving of restored editors to the current context for now.

adietish commented:
@lstocchi this can be tested, basics work fine afaics. Unit tests need to be fixed (I broke the API) and further ones added for the new components.
One can edit & save the kubernetes and the custom types. 
* editors use schemas if they're available (not the case for custom resources):
  ![image](https://user-images.githubusercontent.com/25126/116880905-e2457380-ac22-11eb-934d-40e48e92dcbf.png)
* when saving, the user gets asked to confirm:
  ![image](https://user-images.githubusercontent.com/25126/116881754-f047c400-ac23-11eb-8323-762f9e96d9ec.png)
* If the resource that you're editing was deleted on the cluster, you'll get asked if you want to reload it:
  ![image](https://user-images.githubusercontent.com/25126/116880597-7e22af80-ac22-11eb-9113-31f1901cf4fc.png)
  Ideally one would be able to immediately save the editor back to the cluster. For now one has to edit the content before one can save it (#202)

adietish commented:
@lstocchi changing resource name and namespace doesn't work properly yet. Needs more work. Other changes should work.

adietish commented:
@lstocchi changing name, namespace or kind works now: a new resource is created instead of trying to replace the existing one (which then would fail of course)

lstocchi commented:
@adietish I noted that if you select multiple tree items the context menu shows the `edit` command but this should only be visible when a single item is selected, no?

adietish commented:
> @adietish I noted that if you select multiple tree items the context menu shows the `edit` command but this should only be visible when a single item is selected, no?

you're right. Will restrict to single selection. On the long run, it should be possible to open several editors at once. Filing an enhancement

adietish commented:
@lstocchi, @jeffmaury: made sure that one can completely replace the content of the editor (ex. start editing with a pod, replace with the yaml for a bc) works. What's not annoying is that the editor title should reflect the change, keeping the original title is misleading.

adietish commented:
@lstocchi, @jeffmaury : if a new resource is created (editor content changed so that resource-name, -namespace, -apiVersion, -kind are different), the virtual file that holds the yaml is now renamed, which causes the editor to change the title.

adietish commented:
testing this extensively showed that the "Do you want to save to the cluster?" dialog was coming up far too often (ex. when switching to another app, browser etc.) and is annoying because one has to confirm/refute before one can continue to work. 
![image](https://user-images.githubusercontent.com/25126/119145654-25d51580-ba4a-11eb-95dc-adb7ac3b241d.png)
It turns out that IJ doesn't know the concept of explicit saving like one knows from Eclipse and other editors. It does regular autosaving instead. Explicit saving of an editor is thus the exception, not the rule. One can overturn this behavior in the preferences though, but I'd think that the majority of the users doesn't:
![image](https://user-images.githubusercontent.com/25126/119145374-dc84c600-ba49-11eb-995f-d1b1ba4c2be0.png)
I discussed this with @lstocchi and we thought that we could replace the dialog by a notification/banner in the editor that allows the user to save to the cluster. 
![image](https://user-images.githubusercontent.com/25126/119145898-63d23980-ba4a-11eb-8f5a-b74f573cc9a3.png)
(Do you want to push to the cluster? "Push")
This would be explicit and thus intuitive to the novice user and non-intrusive since one can simply ignore the banner and continue editing.

adietish commented:
@lstocchi, @jeffmaury: ready for testing, please review.
I dont use the save confirmation dialog any more. The editor shows notifications when changes to the resource happen (modified on server, deleted on server, local update etc). In those notifications actions are suggested: push to cluster, reload from cluster.
I added the numerous usecases to `testing-editor.md`:
```
**Editor supports schema**
1. "Edit..." resource

-> editor has corresponding schema selected (bottom right combo "Schema:")

**Editor title is <resource-name>@<namespace-name>.yml**
1. "Edit..." resource

-> editor has title matching pattern <resource>@<namespace>.yml

**Editor file will rename file to <resource>@<namespace>(2).yml if name already used**
1. "Edit..." resource
1. "Edit..." other resource
1. editor 1: Ctrl + A, Ctrl + C
1. editor 2: Ctrl + v

-> editor has title matching pattern <resource-name>@<namespace-name>(2).yml

**Irrelevant change, no notification**
1. "Edit..." resource
1. add space after property value

-> no notification

**Push notification with "update existing"**
1. "Edit..." resource
1. add label
   -> push notification ("Update" not "Create")
1. hit "Push"

-> notification disappears
-> label was added
-> editor title unchanged

**Push notification with "create new"**
1. "Edit..." resource
1. change metadata > name / metadata > namespace / kind
   -> push notification ("create new" not "update existing")
1. hit "Push"

-> new resource appears in tree
-> editor title changes

**Push notification with "update existing" turns "create new"**
1. "Edit..." resource
1. add label
   -> push notification "update existing" (not "create")
1. change metadata > name

-> push notification with "create new"

**Deleted notification appears**
1. "Edit..." resource
1. delete resource (ctx action, console, kubectl)

-> deleted notification appears

**Modified notification appears**
1. "Edit..." resource
1. modify resource externally (console, kubectl)

-> modified notification appears

**Error notification appears when pasting invalid content**
1. "Edit..." resource
2. paste invalid yaml

**Error notification appears on startup**
1. have editor with invalid content
2. restart IJ

-> Error notification appears

**Error notification disappears when correcting invalid content**
1. have editor with invalid content
2. correct error/paste valid content

**Push notification with reload appears**
1. "Edit..." resource
1. modify resource (ex. change label)
   -> Push notification appears
1. modify resource externally

-> Push notification shows additional option "Reload"

**Reload -> push notification disappears**
1. "Edit..." resource
1. modify resource (ex. change label)
   -> Push notification appears
1. modify resource externally
   -> Push notification shows additional option "Reload"
1. hit "Reload"

-> editor shows new resource, notification disappears

**Push now -> push notification disappears**
1. "Edit..." resource
1. modify resource (ex. change label)
   -> Push notification appears
1. modify resource externally
   -> Push notification shows additional option "Reload"
1. hit "Push now"

-> editor stays the same, notification disappears
```

adietish commented:
Unit tests need to be fixed and added though. Extensive manual testing nevertheless very welcome.

lstocchi reviewed (commented):
@adietish i'll keep playing with it as this is huge and it requires some testing but I want to share my findings so far so you can start thinking/working on them if you want. 
The first impression is really good. Having the message in the toolbar with the command to push is simple and easy to understand. So GJ :rocket: 

I have found some misbehavior, not sure if you are aware of them.
1) If I change the kind or some other properties which is not allowed, the toolbar notify it (:+1: ) but then it is stuck and i cannot do anything else. The only option is to close the editor and reopen it.

To replicate it:

- Open a pod resource
- Change the kind so that it is invalid (e.g Posd)
- The details action shows an empty field and the tollbar seems stuck

![pods_error](https://user-images.githubusercontent.com/49404737/119829512-9b822b00-befb-11eb-8285-33f68ac8546c.gif)

2) The second problem is if i open a yaml file in IJ which contains a kubernetes resource. If i start editing it, the plugin changes its name (to update the name shown in the editor tab i guess). If i'm not aware of it i don't find it anymore in my filesystem. (e.g i have a file called mydeployment.yaml which contains a deployment with name = blabla, i open it in IJ,  i start editing it, the file is renamed blabla.yml. At this point if i close IJ and start looking for my original file i don't find it and i'm lost).

3) The third is just a tiny thing. I saw that with custom resources (everything from tekton/knative in my case) the editor opens up with the `push to cluster` toolbar activated by default. Any idea why?
![push_present](https://user-images.githubusercontent.com/49404737/119830944-097b2200-befd-11eb-8353-72cbf8c0c875.gif)

jeffmaury commented:
I have playing with the knativetutorial so I wanted to edit my knative service to I went to custom resources ->services and select edit but got:

```
java.lang.IllegalStateException: javaClass.getResourceAsS…, INDEX_FILE).toString()) must not be null
	at com.redhat.devtools.intellij.kubernetes.validation.KubernetesSchemasProviderFactory.load(KubernetesSchemaProviderFactory.kt:49)
	at com.redhat.devtools.intellij.kubernetes.validation.KubernetesSchemasProviderFactory.getProviders(KubernetesSchemaProviderFactory.kt:38)
	at com.redhat.devtools.intellij.kubernetes.validation.KubernetesSchemasProviderFactory.getProviders(KubernetesSchemaProviderFactory.kt:45)
	at com.jetbrains.jsonSchema.impl.JsonSchemaServiceImpl.getProvidersFromFactories(JsonSchemaServiceImpl.java:77)
	at com.jetbrains.jsonSchema.impl.JsonSchemaServiceImpl.lambda$new$0(JsonSchemaServiceImpl.java:47)
	at com.jetbrains.jsonSchema.impl.JsonSchemaServiceImpl$MyState$1.compute(JsonSchemaServiceImpl.java:421)
	at com.jetbrains.jsonSchema.impl.JsonSchemaServiceImpl$MyState$1.compute(JsonSchemaServiceImpl.java:416)
	at com.intellij.openapi.util.ClearableLazyValue.getValue(ClearableLazyValue.java:48)
	at com.jetbrains.jsonSchema.impl.JsonSchemaServiceImpl$MyState$1.getValue(JsonSchemaServiceImpl.java:427)
	at com.jetbrains.jsonSchema.impl.JsonSchemaServiceImpl$MyState$1.getValue(JsonSchemaServiceImpl.java:416)
	at com.jetbrains.jsonSchema.impl.JsonSchemaServiceImpl$MyState.getProviders(JsonSchemaServiceImpl.java:444)
	at com.jetbrains.jsonSchema.impl.JsonSchemaServiceImpl.getProvidersForFile(JsonSchemaServiceImpl.java:201)
	at com.jetbrains.jsonSchema.impl.JsonSchemaConflictNotificationProvider.hasConflicts(JsonSchemaConflictNotificationProvider.java:84)
	at com.jetbrains.jsonSchema.impl.JsonSchemaConflictNotificationProvider.createNotificationPanel(JsonSchemaConflictNotificationProvider.java:68)
	at com.jetbrains.jsonSchema.impl.JsonSchemaConflictNotificationProvider.createNotificationPanel(JsonSchemaConflictNotificationProvider.java:43)
	at com.intellij.ui.EditorNotificationsImpl$3.performInReadAction(EditorNotificationsImpl.java:136)
	at com.intellij.openapi.progress.util.ReadTask.lambda$runBackgroundProcess$0(ReadTask.java:66)
	at com.intellij.openapi.application.impl.ApplicationImpl.runReadAction(ApplicationImpl.java:977)
	at com.intellij.openapi.application.ReadAction.compute(ReadAction.java:57)
	at com.intellij.openapi.progress.util.ReadTask.runBackgroundProcess(ReadTask.java:66)
	at com.intellij.openapi.progress.util.ProgressIndicatorUtils.lambda$runUnderProgress$4(ProgressIndicatorUtils.java:227)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcess$3(CoreProgressManager.java:180)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcess$2(CoreProgressManager.java:164)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:582)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:532)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:87)
	at com.intellij.openapi.progress.impl.CoreProgressManager.runProcess(CoreProgressManager.java:151)
	at com.intellij.openapi.progress.impl.CoreProgressManager.runProcess(CoreProgressManager.java:180)
	at com.intellij.openapi.progress.util.ProgressIndicatorUtils.runUnderProgress(ProgressIndicatorUtils.java:225)
	at com.intellij.openapi.progress.util.ProgressIndicatorUtils.access$100(ProgressIndicatorUtils.java:33)
	at com.intellij.openapi.progress.util.ProgressIndicatorUtils$4.run(ProgressIndicatorUtils.java:176)
	at com.intellij.util.concurrency.BoundedTaskExecutor.doRun(BoundedTaskExecutor.java:227)
	at com.intellij.util.concurrency.BoundedTaskExecutor.access$100(BoundedTaskExecutor.java:26)
	at com.intellij.util.concurrency.BoundedTaskExecutor$2$1.run(BoundedTaskExecutor.java:200)
	at com.intellij.util.ConcurrencyUtil.runUnderThreadName(ConcurrencyUtil.java:229)
	at com.intellij.util.concurrency.BoundedTaskExecutor$2.run(BoundedTaskExecutor.java:194)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

adietish commented:
@jeffmaury 
> I have playing with the knativetutorial 
Do you have some links or instructions to share so that I can try to replicate this?

adietish commented:
> The first impression is really good. Having the message in the toolbar with the command to push is simple and easy to understand. So GJ 🚀

awesome, thanks for this feedback. Confirms to me that this approach is worth it.

>     1. The second problem is if i open a yaml file in IJ which contains a kubernetes resource. If i start editing it
How would you "open a yaml file"? For now the workflow starts with you editing a resource in the tree?
Do you have some steps for me to understand your usecase please?
Btw. you're right, I'm renaming the file, that's the only way I found to have the editor title changing.

> 
>     2. The third is just a tiny thing. I saw that with custom resources (everything from tekton/knative in my case) the editor opens up with the `push to cluster` toolbar activated by default. Any idea why?
That's interesting, unexpected. The "push" notification appears if your local file is modified when compared to the remote resource. If the remote had changed too, you'd also have the option in the notification to load the resource from the cluster. Apparently that's not the case here. Do you have any insight why this is the case? Maybe that's a file that wasn't deleted when you were editing before?

jeffmaury commented:
> 
> 
> @jeffmaury
> 
> > I have playing with the knativetutorial
> 
> Do you have some links or instructions to share so that I can try to replicate this?

https://redhat-developer-demos.github.io/knative-tutorial/knative-tutorial/index.html

adietish commented:
@lstocchi: 
>     1. The second problem is if i open a yaml file in IJ which contains a kubernetes resource. If i start editing it, the plugin changes its name (to update the name shown in the editor tab i guess). 

I disabled the editor to deal with files that you load from your disk via File > Open. I created #217 to deal with this once this PR is merged

adietish commented:
> I have playing with the knativetutorial so I wanted to edit my knative service to I went to custom resources ->services and select edit but got:
> 
> ```
> java.lang.IllegalStateException: javaClass.getResourceAsS…, INDEX_FILE).toString()) must not be null
> 	at com.redhat.devtools.intellij.kubernetes.validation.KubernetesSchemasProviderFactory.load(KubernetesSchemaProviderFactory.kt:49)
> 	at com.redhat.devtools.intellij.kubernetes.validation.KubernetesSchemasProviderFactory.getProviders(KubernetesSchemaProviderFactory.kt:38)
> ```

I cannot reproduce this. I suspect that this error should go away if you clean before building?

> 
> Also when I want to push the changes, then I got:
> 
> ![image](https://user-images.githubusercontent.com/695993/119867873-a5695580-bf1e-11eb-9582-8b7ebdb8e744.png)

I can replicate this. Fixing it.

> Also got the same error while trying to edit a pod

Could not replicate this.

> 
> Also after I closed the editor, it seems the edited filed is not deleted as I have (when I reopen the editor):
> 
> ![image](https://user-images.githubusercontent.com/695993/119868033-d34e9a00-bf1e-11eb-95ba-2b07fd3a46f7.png)

This should be fixed.

adietish commented:
@lstocchi:

> @adietish i'll keep playing with it as this is huge and it requires some testing but I want to share my findings so far so you can start thinking/working on them if you want.
> The first impression is really good. Having the message in the toolbar with the command to push is simple and easy to understand. So GJ 🚀
> 
> I have found some misbehavior, not sure if you are aware of them.
> 
>     1. If I change the kind or some other properties which is not allowed, the toolbar notify it (👍 ) but then it is stuck and i cannot do anything else. The only option is to close the editor and reopen it.
> 

This is fixed.

>     1. The second problem is if i open a yaml file in IJ which contains a kubernetes resource. If i start editing it, the plugin changes its name (to update the name shown in the editor tab i guess). If i'm not aware of it i don't find it anymore in my filesystem. (e.g i have a file called mydeployment.yaml which contains a deployment with name = blabla, i open it in IJ,  i start editing it, the file is renamed blabla.yml. 

I fixed this by disabling "my" notifications in the "normal" yaml editors. This should be reintroduced correctly in #217 

> 
>     2. The third is just a tiny thing. I saw that with custom resources (everything from tekton/knative in my case) the editor opens up with the `push to cluster` toolbar activated by default. Any idea why?
>        ![push_present](https://user-images.githubusercontent.com/49404737/119830944-097b2200-befd-11eb-8353-72cbf8c0c875.gif)

This is fixed.

adietish commented:
@jeffmaury:

> 
> > I have playing with the knativetutorial so I wanted to edit my knative service to I went to custom resources ->services and select edit but got:
> > ```
> > java.lang.IllegalStateException: javaClass.getResourceAsS…, INDEX_FILE).toString()) must not be null
> > 	at com.redhat.devtools.intellij.kubernetes.validation.KubernetesSchemasProviderFactory.load(KubernetesSchemaProviderFactory.kt:49)
> > 	at com.redhat.devtools.intellij.kubernetes.validation.KubernetesSchemasProviderFactory.getProviders(KubernetesSchemaProviderFactory.kt:38)
> > ```
> 
> ❓ I cannot reproduce this. I suspect that this error should go away if you clean before building?
> 
> > Also when I want to push the changes, then I got:
> > ![image](https://user-images.githubusercontent.com/695993/119867873-a5695580-bf1e-11eb-9582-8b7ebdb8e744.png)
> 
> ✅ This should be fixed.
> 
> > Also got the same error while trying to edit a pod
> 
> ❓ Could not replicate this. Can you please report the resource apiVersion that is reported as not supported?
> 
> > Also after I closed the editor, it seems the edited filed is not deleted as I have (when I reopen the editor):
> > ![image](https://user-images.githubusercontent.com/695993/119868033-d34e9a00-bf1e-11eb-95ba-2b07fd3a46f7.png)
> 
> ✅ This should be fixed.

Please test again.

adietish commented:
@lstocchi, @jeffmaury, @olkornii: this is ready for review. All usecases in the document are passing for me.

lstocchi reviewed (approved):
LGTM. Impressive work.

Just a comment below and a possible idea for future enhancements.

When trying to edit a resource it may happen that the notification bar doesn't popup (for some reason) and users don't know what to do. Maybe we could give them a way to "force" a push, maybe with a command in the context menu of the editor(??).

A way I found to replicate it is:
Open a tekton task with `apiVersion: v1beta1`.
Change it to `apiVersion: v1alpha1`
Edit the spec
No notification bar pop up and i don't know what to do if i want to push it anyway.

adietish commented:
> LGTM. Impressive work.
> 
> Just a comment below and a possible idea for future enhancements.
> 
> When trying to edit a resource it may happen that the notification bar doesn't popup (for some reason) and users don't know what to do. Maybe we could give them a way to "force" a push, maybe with a command in the context menu of the editor(??).

great point. :thumbsup:
filed #218 

> 
> A way I found to replicate it is:
> Open a tekton task with `apiVersion: v1beta1`.
> Change it to `apiVersion: v1alpha1`
> Edit the spec
> No notification bar pop up and i don't know what to do if i want to push it anyway.

thanks for reporting this. I'll file an issue and fix it.

adietish commented:
@jeffmaury refactoring is finished. Please review

jeffmaury commented:
Still having some issues. I had one pod from my knative setup, asked for edit and got this error reported by IJ:

2021-06-14 10:38:16,077 [ 104762]  ERROR - onSchema.ide.JsonSchemaService - javaClass.getResourceAsS…, INDEX_FILE).toString()) must not be null 
java.lang.IllegalStateException: javaClass.getResourceAsS…, INDEX_FILE).toString()) must not be null
	at com.redhat.devtools.intellij.kubernetes.validation.KubernetesSchemasProviderFactory.load(KubernetesSchemaProviderFactory.kt:49)
	at com.redhat.devtools.intellij.kubernetes.validation.KubernetesSchemasProviderFactory.getProviders(KubernetesSchemaProviderFactory.kt:38)
	at com.redhat.devtools.intellij.kubernetes.validation.KubernetesSchemasProviderFactory.getProviders(KubernetesSchemaProviderFactory.kt:45)
	at com.jetbrains.jsonSchema.impl.JsonSchemaServiceImpl.getProvidersFromFactories(JsonSchemaServiceImpl.java:77)
	at com.jetbrains.jsonSchema.impl.JsonSchemaServiceImpl.lambda$new$0(JsonSchemaServiceImpl.java:47)
	at com.jetbrains.jsonSchema.impl.JsonSchemaServiceImpl$MyState$1.compute(JsonSchemaServiceImpl.java:421)
	at com.jetbrains.jsonSchema.impl.JsonSchemaServiceImpl$MyState$1.compute(JsonSchemaServiceImpl.java:416)
	at com.intellij.openapi.util.ClearableLazyValue.getValue(ClearableLazyValue.java:48)
	at com.jetbrains.jsonSchema.impl.JsonSchemaServiceImpl$MyState$1.getValue(JsonSchemaServiceImpl.java:427)
	at com.jetbrains.jsonSchema.impl.JsonSchemaServiceImpl$MyState$1.getValue(JsonSchemaServiceImpl.java:416)
	at com.jetbrains.jsonSchema.impl.JsonSchemaServiceImpl$MyState.getProviders(JsonSchemaServiceImpl.java:444)
	at com.jetbrains.jsonSchema.impl.JsonSchemaServiceImpl.getProvidersForFile(JsonSchemaServiceImpl.java:201)
	at com.jetbrains.jsonSchema.impl.JsonSchemaServiceImpl.getSchemasForFile(JsonSchemaServiceImpl.java:145)
	at com.jetbrains.jsonSchema.impl.JsonSchemaServiceImpl.getSchemaFilesForFile(JsonSchemaServiceImpl.java:129)
	at com.jetbrains.jsonSchema.widget.JsonSchemaStatusWidget.getWidgetState(JsonSchemaStatusWidget.java:96)
	at com.intellij.openapi.wm.impl.status.EditorBasedStatusBarPopup.lambda$update$0(EditorBasedStatusBarPopup.java:207)
	at com.intellij.util.concurrency.QueueProcessor.runSafely(QueueProcessor.java:232)
	at com.intellij.util.Alarm$Request.runSafely(Alarm.java:356)
	at com.intellij.util.Alarm$Request.run(Alarm.java:343)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at com.intellij.util.concurrency.SchedulingWrapper$MyScheduledFutureTask.run(SchedulingWrapper.java:228)
	at com.intellij.openapi.application.impl.LaterInvocator$FlushQueue.doRun(LaterInvocator.java:435)
	at com.intellij.openapi.application.impl.LaterInvocator$FlushQueue.runNextEvent(LaterInvocator.java:419)
	at com.intellij.openapi.application.impl.LaterInvocator$FlushQueue.run(LaterInvocator.java:403)
	at java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:311)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:762)
	at java.awt.EventQueue.access$500(EventQueue.java:98)
	at java.awt.EventQueue$3.run(EventQueue.java:715)
	at java.awt.EventQueue$3.run(EventQueue.java:709)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:80)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:732)
	at com.intellij.ide.IdeEventQueue.defaultDispatchEvent(IdeEventQueue.java:719)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.java:668)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:363)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:201)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)

adietish commented:
That's odd. I can't replicate this. According to the stack trace it fails to load the index file that holds all files that were generated:

> javaClass.getResourceAsS…, INDEX_FILE).toString()) must not be null

The file is at `resources/schemas.k8s.io/index.txt`. Here's my content:
```
cindervolumesource-v1.json
volumenoderesources-storage-v1beta1.json
ingressstatus-networking-v1beta1.json
poddisruptionbudgetlist-policy-v1beta1.json
endpointslist-v1.json
eviction-policy-v1beta1.json
queuingconfiguration-flowcontrol-v1alpha1.json
policyrule-rbac-v1alpha1.json
poddnsconfigoption-v1.json
podcondition-v1.json
rolebinding-rbac-v1alpha1.json
scheduling-node-v1.json
objectreference-v1.json
crossversionobjectreference-autoscaling-v2beta1.json
ingressspec-networking-v1beta1.json
containerresourcemetricsource-autoscaling-v2beta2.json
nfsvolumesource-v1.json
secretenvsource-v1.json
runasuserstrategyoptions-policy-v1beta1.json
ingresstls-extensions-v1beta1.json
ingressbackend-extensions-v1beta1.json
externalmetricstatus-autoscaling-v2beta2.json
apiservicespec-apiregistration-v1.json
rolelist-rbac-v1.json
attachedvolume-v1.json
tokenreviewspec-authentication-v1beta1.json
flowschemastatus-flowcontrol-v1beta1.json
clusterrole-rbac-v1.json
customresourcedefinitionnames-apiextensions-v1.json
role-rbac-v1beta1.json
horizontalpodautoscalercondition-autoscaling-v2beta1.json
metricstatus-autoscaling-v2beta2.json
json-apiextensions-v1.json
objectmetricstatus-autoscaling-v2beta2.json
metricvaluestatus-autoscaling-v2beta2.json
podsecuritycontext-v1.json
apiservice-apiregistration-v1.json
toleration-v1.json
roleref-rbac-v1beta1.json
volumeattachmentlist-storage-v1.json
mutatingwebhookconfiguration-admissionregistration-v1beta1.json
horizontalpodautoscalerstatus-autoscaling-v2beta2.json
containerstatewaiting-v1.json
glusterfspersistentvolumesource-v1.json
persistentvolumeclaimvolumesource-v1.json
ingressclassspec-networking-v1beta1.json
overhead-node-v1beta1.json
rolebindinglist-rbac-v1alpha1.json
deployment-apps-v1.json
volumeattachmentlist-storage-v1beta1.json
nodesysteminfo-v1.json
scalespec-autoscaling-v1.json
metricspec-autoscaling-v2beta2.json
statefulset-apps-v1.json
podtemplatespec-v1.json
azurefilevolumesource-v1.json
allowedhostpath-policy-v1beta1.json
volumeattachmentsource-storage-v1.json
serviceaccountlist-v1.json
componentstatuslist-v1.json
componentstatus-v1.json
selfsubjectaccessreview-authorization-v1.json
subject-flowcontrol-v1beta1.json
overhead-node-v1alpha1.json
ingresslist-networking-v1.json
portstatus-v1.json
resourcequotaspec-v1.json
deploymentstatus-apps-v1.json
servicereference-apiregistration-v1beta1.json
tokenrequestspec-authentication-v1.json
limitedprioritylevelconfiguration-flowcontrol-v1beta1.json
podtemplatelist-v1.json
allowedcsidriver-policy-v1beta1.json
podstatus-v1.json
joblist-batch-v1.json
validatingwebhook-admissionregistration-v1.json
nodedaemonendpoints-v1.json
clusterrolelist-rbac-v1beta1.json
apiservicestatus-apiregistration-v1beta1.json
poddisruptionbudget-policy-v1beta1.json
customresourcesubresources-apiextensions-v1.json
runasgroupstrategyoptions-policy-v1beta1.json
persistentvolumeclaimtemplate-v1.json
preferredschedulingterm-v1.json
resourcequotalist-v1.json
resourcemetricsource-autoscaling-v2beta2.json
job-batch-v1.json
ingressclassspec-networking-v1.json
volumeattachmentstatus-storage-v1.json
idrange-policy-v1beta1.json
priorityclass-scheduling-v1alpha1.json
nodeconfigstatus-v1.json
hostportrange-policy-v1beta1.json
runtimeclasslist-node-v1beta1.json
validatingwebhookconfigurationlist-admissionregistration-v1.json
horizontalpodautoscalerlist-autoscaling-v1.json
selfsubjectrulesreview-authorization-v1beta1.json
nodeconfigsource-v1.json
customresourcesubresources-apiextensions-v1beta1.json
volumeattachment-storage-v1alpha1.json
daemonendpoint-v1.json
csidriverlist-storage-v1.json
customresourcesubresourcestatus-apiextensions-v1.json
iscsipersistentvolumesource-v1.json
apiservicestatus-apiregistration-v1.json
mutatingwebhookconfiguration-admissionregistration-v1.json
statefulsetstatus-apps-v1.json
volumeattachmentspec-storage-v1beta1.json
statuscause-meta-v1.json
mutatingwebhookconfigurationlist-admissionregistration-v1beta1.json
clusterrolebindinglist-rbac-v1beta1.json
horizontalpodautoscalerlist-autoscaling-v2beta2.json
crossversionobjectreference-autoscaling-v1.json
networkpolicypeer-networking-v1.json
podtemplate-v1.json
emptydirvolumesource-v1.json
runtimeclasslist-node-v1.json
csidriver-storage-v1beta1.json
daemonsetcondition-apps-v1.json
labelselectorrequirement-meta-v1.json
nodecondition-v1.json
overhead-node-v1.json
container-v1.json
resourcefieldselector-v1.json
persistentvolumeclaimlist-v1.json
flowschemaspec-flowcontrol-v1beta1.json
selfsubjectrulesreviewspec-authorization-v1beta1.json
secretkeyselector-v1.json
customresourcesubresourcescale-apiextensions-v1.json
volumemount-v1.json
capabilities-v1.json
rollingupdatedaemonset-apps-v1.json
fcvolumesource-v1.json
gitrepovolumesource-v1.json
selfsubjectaccessreviewspec-authorization-v1beta1.json
customresourceconversion-apiextensions-v1beta1.json
typedlocalobjectreference-v1.json
topologyselectorlabelrequirement-v1.json
namespace-v1.json
csinodelist-storage-v1.json
ingresslist-extensions-v1beta1.json
validatingwebhookconfiguration-admissionregistration-v1beta1.json
csipersistentvolumesource-v1.json
ingress-networking-v1.json
persistentvolumeclaimspec-v1.json
flowschemaspec-flowcontrol-v1alpha1.json
replicasetspec-apps-v1.json
quantity-resource.json
topologyselectorterm-v1.json
podsmetricsource-autoscaling-v2beta2.json
event-events-v1beta1.json
affinity-v1.json
prioritylevelconfigurationreference-flowcontrol-v1alpha1.json
cronjobspec-batch-v1beta1.json
topologyspreadconstraint-v1.json
cinderpersistentvolumesource-v1.json
priorityclass-scheduling-v1.json
cronjobstatus-batch-v2alpha1.json
csidriverlist-storage-v1beta1.json
storageclasslist-storage-v1beta1.json
networkpolicy-networking-v1.json
customresourcedefinitionnames-apiextensions-v1beta1.json
leasespec-coordination-v1beta1.json
httpingresspath-networking-v1beta1.json
nodeselectorterm-v1.json
certificatesigningrequestlist-certificates-v1beta1.json
ingressbackend-networking-v1.json
clusterrolebindinglist-rbac-v1alpha1.json
csinodedriver-storage-v1.json
ingresstls-networking-v1.json
limitrange-v1.json
volumeattachmentspec-storage-v1.json
limitrangespec-v1.json
rollingupdatedeployment-apps-v1.json
replicaset-apps-v1.json
windowssecuritycontextoptions-v1.json
resourcequota-v1.json
certificatesigningrequest-certificates-v1.json
statefulsetlist-apps-v1.json
nodeselector-v1.json
ipblock-networking-v1.json
horizontalpodautoscalerspec-autoscaling-v1.json
endpointsubset-v1.json
rolebindinglist-rbac-v1beta1.json
volumeerror-storage-v1beta1.json
configmapkeyselector-v1.json
priorityclasslist-scheduling-v1.json
nodeaddress-v1.json
resourcemetricstatus-autoscaling-v2beta1.json
volumeprojection-v1.json
eventlist-events-v1.json
servicebackendport-networking-v1.json
sessionaffinityconfig-v1.json
limitrangelist-v1.json
horizontalpodautoscaler-autoscaling-v2beta1.json
eventseries-events-v1.json
httpingressrulevalue-extensions-v1beta1.json
gcepersistentdiskvolumesource-v1.json
runtimeclassspec-node-v1alpha1.json
allowedflexvolume-policy-v1beta1.json
subjectaccessreviewspec-authorization-v1beta1.json
externalmetricsource-autoscaling-v2beta1.json
service-v1.json
storageversionspec-apiserverinternal-v1alpha1.json
ownerreference-meta-v1.json
containerresourcemetricstatus-autoscaling-v2beta1.json
serverstorageversion-apiserverinternal-v1alpha1.json
azurediskvolumesource-v1.json
info-pkg-version.json
prioritylevelconfigurationspec-flowcontrol-v1beta1.json
nodespec-v1.json
volumeattachment-storage-v1.json
webhookclientconfig-admissionregistration-v1.json
projectedvolumesource-v1.json
daemonsetlist-apps-v1.json
runtimeclass-node-v1alpha1.json
selfsubjectaccessreviewspec-authorization-v1.json
objectmetricsource-autoscaling-v2beta1.json
userinfo-authentication-v1beta1.json
taint-v1.json
scaleiopersistentvolumesource-v1.json
event-events-v1.json
microtime-meta-v1.json
nonresourcepolicyrule-flowcontrol-v1beta1.json
horizontalpodautoscalerstatus-autoscaling-v1.json
persistentvolumeclaimcondition-v1.json
certificatesigningrequest-certificates-v1beta1.json
policyrule-rbac-v1beta1.json
networkpolicylist-networking-v1.json
flexpersistentvolumesource-v1.json
scheduling-node-v1alpha1.json
nodelist-v1.json
containerstateterminated-v1.json
tokenrequest-storage-v1.json
aggregationrule-rbac-v1.json
probe-v1.json
rbdpersistentvolumesource-v1.json
policyrule-rbac-v1.json
secretreference-v1.json
index.txt
resourcerule-authorization-v1.json
volumeattachmentstatus-storage-v1alpha1.json
volumeerror-storage-v1alpha1.json
horizontalpodautoscalerspec-autoscaling-v2beta1.json
rolelist-rbac-v1alpha1.json
flowschemastatus-flowcontrol-v1alpha1.json
metricidentifier-autoscaling-v2beta2.json
prioritylevelconfigurationlist-flowcontrol-v1alpha1.json
scale-autoscaling-v1.json
storageclass-storage-v1beta1.json
userinfo-authentication-v1.json
metrictarget-autoscaling-v2beta2.json
apiservicecondition-apiregistration-v1beta1.json
servicestatus-v1.json
endpointconditions-discovery-v1beta1.json
flexvolumesource-v1.json
cronjob-batch-v2alpha1.json
tokenreviewstatus-authentication-v1beta1.json
role-rbac-v1.json
podsmetricstatus-autoscaling-v2beta1.json
priorityclasslist-scheduling-v1alpha1.json
role-rbac-v1alpha1.json
configmap-v1.json
servicereference-admissionregistration-v1.json
deploymentspec-apps-v1.json
persistentvolumeclaim-v1.json
event-v1.json
scheduling-node-v1beta1.json
ingressrule-networking-v1.json
ingress-extensions-v1beta1.json
patch-meta-v1.json
apiversions-meta-v1.json
serviceaccountsubject-flowcontrol-v1beta1.json
csidriverspec-storage-v1.json
flowdistinguishermethod-flowcontrol-v1alpha1.json
volumeattachment-storage-v1beta1.json
serviceaccounttokenprojection-v1.json
podaffinity-v1.json
resourcerequirements-v1.json
poddisruptionbudgetspec-policy-v1beta1.json
replicationcontrollerstatus-v1.json
rulewithoperations-admissionregistration-v1beta1.json
ingressrule-extensions-v1beta1.json
localvolumesource-v1.json
rawextension-pkg-runtime.json
cephfsvolumesource-v1.json
prioritylevelconfigurationcondition-flowcontrol-v1beta1.json
httpingresspath-networking-v1.json
serviceaccountsubject-flowcontrol-v1alpha1.json
ingressclass-networking-v1.json
jobtemplatespec-batch-v1beta1.json
subjectaccessreviewstatus-authorization-v1.json
containerstate-v1.json
clusterrolebinding-rbac-v1beta1.json
tcpsocketaction-v1.json
apiservicelist-apiregistration-v1beta1.json
containerimage-v1.json
subjectaccessreview-authorization-v1beta1.json
containerport-v1.json
runtimeclass-node-v1.json
time-meta-v1.json
managedfieldsentry-meta-v1.json
roleref-rbac-v1.json
jobstatus-batch-v1.json
deleteoptions-meta-v1.json
rolelist-rbac-v1beta1.json
_definitions.json
controllerrevisionlist-apps-v1.json
volume-v1.json
csidriverspec-storage-v1beta1.json
horizontalpodautoscalerbehavior-autoscaling-v2beta2.json
csinodespec-storage-v1.json
deploymentcondition-apps-v1.json
customresourcedefinitioncondition-apiextensions-v1beta1.json
aggregationrule-rbac-v1beta1.json
servicereference-apiextensions-v1.json
storageversionlist-apiserverinternal-v1alpha1.json
controllerrevision-apps-v1.json
poddisruptionbudgetstatus-policy-v1beta1.json
prioritylevelconfiguration-flowcontrol-v1beta1.json
rulewithoperations-admissionregistration-v1.json
rollingupdatestatefulsetstrategy-apps-v1.json
supplementalgroupsstrategyoptions-policy-v1beta1.json
networkpolicyport-networking-v1.json
cronjobstatus-batch-v1beta1.json
eventsource-v1.json
groupversionfordiscovery-meta-v1.json
webhookconversion-apiextensions-v1.json
watchevent-meta-v1.json
podreadinessgate-v1.json
mutatingwebhookconfigurationlist-admissionregistration-v1.json
daemonsetspec-apps-v1.json
clusterrolebinding-rbac-v1.json
configmaplist-v1.json
daemonset-apps-v1.json
networkpolicyegressrule-networking-v1.json
poddnsconfig-v1.json
ingressrule-networking-v1beta1.json
clusterrolebinding-rbac-v1alpha1.json
mutatingwebhook-admissionregistration-v1.json
eventlist-v1.json
rolebinding-rbac-v1beta1.json
cronjoblist-batch-v2alpha1.json
ingress-networking-v1beta1.json
resourcepolicyrule-flowcontrol-v1alpha1.json
rolebinding-rbac-v1.json
customresourcecolumndefinition-apiextensions-v1.json
namespacecondition-v1.json
flowschemalist-flowcontrol-v1beta1.json
csinodedriver-storage-v1beta1.json
groupsubject-flowcontrol-v1beta1.json
leaselist-coordination-v1.json
tokenreviewstatus-authentication-v1.json
customresourceconversion-apiextensions-v1.json
eventseries-v1.json
cephfspersistentvolumesource-v1.json
externaldocumentation-apiextensions-v1beta1.json
statefulsetupdatestrategy-apps-v1.json
aggregationrule-rbac-v1alpha1.json
nodeaffinity-v1.json
eventlist-events-v1beta1.json
limitedprioritylevelconfiguration-flowcontrol-v1alpha1.json
serviceport-v1.json
portworxvolumesource-v1.json
statefulsetspec-apps-v1.json
fsgroupstrategyoptions-policy-v1beta1.json
runtimeclasslist-node-v1alpha1.json
certificatesigningrequestlist-certificates-v1.json
boundobjectreference-authentication-v1.json
horizontalpodautoscalerspec-autoscaling-v2beta2.json
downwardapivolumefile-v1.json
secretvolumesource-v1.json
apiservicespec-apiregistration-v1beta1.json
flowschema-flowcontrol-v1alpha1.json
runtimeclass-node-v1beta1.json
prioritylevelconfigurationreference-flowcontrol-v1beta1.json
usersubject-flowcontrol-v1beta1.json
podip-v1.json
tokenreview-authentication-v1.json
downwardapiprojection-v1.json
podsmetricstatus-autoscaling-v2beta2.json
configmapvolumesource-v1.json
podantiaffinity-v1.json
ingressstatus-networking-v1.json
localsubjectaccessreview-authorization-v1.json
eventseries-events-v1beta1.json
flowschemalist-flowcontrol-v1alpha1.json
nonresourcepolicyrule-flowcontrol-v1alpha1.json
roleref-rbac-v1alpha1.json
configmapnodeconfigsource-v1.json
resourcerule-authorization-v1beta1.json
statusdetails-meta-v1.json
envvarsource-v1.json
networkpolicyspec-networking-v1.json
nonresourcerule-authorization-v1beta1.json
lease-coordination-v1.json
volumedevice-v1.json
json-apiextensions-v1beta1.json
podspec-v1.json
clusterrole-rbac-v1beta1.json
horizontalpodautoscaler-autoscaling-v2beta2.json
csinode-storage-v1.json
envfromsource-v1.json
networkpolicyingressrule-networking-v1.json
namespacestatus-v1.json
lease-coordination-v1beta1.json
flowschema-flowcontrol-v1beta1.json
selfsubjectaccessreview-authorization-v1beta1.json
hostalias-v1.json
sysctl-v1.json
externalmetricsource-autoscaling-v2beta2.json
clusterrolelist-rbac-v1.json
status-meta-v1.json
flowschemacondition-flowcontrol-v1alpha1.json
volumenodeaffinity-v1.json
containerresourcemetricstatus-autoscaling-v2beta2.json
endpointport-v1.json
configmapprojection-v1.json
scopedresourceselectorrequirement-v1.json
customresourcesubresourcestatus-apiextensions-v1beta1.json
storageversion-apiserverinternal-v1alpha1.json
deploymentstrategy-apps-v1.json
apiresourcelist-meta-v1.json
selfsubjectrulesreview-authorization-v1.json
fieldsv1-meta-v1.json
secret-v1.json
replicationcontrollercondition-v1.json
replicasetstatus-apps-v1.json
flockervolumesource-v1.json
objectmetricsource-autoscaling-v2beta2.json
podlist-v1.json
priorityclasslist-scheduling-v1beta1.json
apiservice-apiregistration-v1beta1.json
certificatesigningrequestspec-certificates-v1.json
listmeta-meta-v1.json
weightedpodaffinityterm-v1.json
storageclass-storage-v1.json
httpheader-v1.json
scopeselector-v1.json
loadbalanceringress-v1.json
volumeattachmentsource-storage-v1alpha1.json
httpingresspath-extensions-v1beta1.json
subject-flowcontrol-v1alpha1.json
pod-v1.json
webhookclientconfig-apiextensions-v1beta1.json
cronjob-batch-v1beta1.json
deploymentlist-apps-v1.json
leasespec-coordination-v1.json
nonresourceattributes-authorization-v1.json
quobytevolumesource-v1.json
customresourcedefinitionstatus-apiextensions-v1.json
certificatesigningrequeststatus-certificates-v1beta1.json
replicationcontrollerlist-v1.json
seccompprofile-v1.json
securitycontext-v1.json
serveraddressbyclientcidr-meta-v1.json
ingressclass-networking-v1beta1.json
rbdvolumesource-v1.json
preconditions-meta-v1.json
hpascalingrules-autoscaling-v2beta2.json
serviceaccount-v1.json
subjectaccessreview-authorization-v1.json
clientipconfig-v1.json
resourcepolicyrule-flowcontrol-v1beta1.json
storageversionstatus-apiserverinternal-v1alpha1.json
ephemeralcontainer-v1.json
subject-rbac-v1alpha1.json
httpingressrulevalue-networking-v1beta1.json
replicationcontrollerspec-v1.json
endpointslice-discovery-v1beta1.json
resourcemetricstatus-autoscaling-v2beta2.json
volumeattachmentlist-storage-v1alpha1.json
objectmeta-meta-v1.json
mutatingwebhook-admissionregistration-v1beta1.json
podsecuritypolicylist-policy-v1beta1.json
clusterrolelist-rbac-v1alpha1.json
hostpathvolumesource-v1.json
resourceattributes-authorization-v1beta1.json
scalestatus-autoscaling-v1.json
validatingwebhookconfigurationlist-admissionregistration-v1beta1.json
envvar-v1.json
tokenrequest-storage-v1beta1.json
prioritylevelconfigurationspec-flowcontrol-v1alpha1.json
subjectaccessreviewspec-authorization-v1.json
downwardapivolumesource-v1.json
photonpersistentdiskvolumesource-v1.json
containerstaterunning-v1.json
customresourcedefinitionstatus-apiextensions-v1beta1.json
secretprojection-v1.json
endpoint-discovery-v1beta1.json
tokenreviewspec-authentication-v1.json
scaleiovolumesource-v1.json
replicationcontroller-v1.json
flowdistinguishermethod-flowcontrol-v1beta1.json
csinodelist-storage-v1beta1.json
horizontalpodautoscaler-autoscaling-v1.json
awselasticblockstorevolumesource-v1.json
runtimeclassstrategyoptions-policy-v1beta1.json
webhookclientconfig-apiextensions-v1.json
hpascalingpolicy-autoscaling-v2beta2.json
policyruleswithsubjects-flowcontrol-v1beta1.json
apiresource-meta-v1.json
podsmetricsource-autoscaling-v2beta1.json
vspherevirtualdiskvolumesource-v1.json
endpointaddress-v1.json
validatingwebhook-admissionregistration-v1beta1.json
podaffinityterm-v1.json
binding-v1.json
tokenrequeststatus-authentication-v1.json
persistentvolumelist-v1.json
subjectrulesreviewstatus-authorization-v1beta1.json
subject-rbac-v1.json
csidriver-storage-v1.json
node-v1.json
certificatesigningrequestcondition-certificates-v1.json
volumeattachmentsource-storage-v1beta1.json
apiservicelist-apiregistration-v1.json
podsecuritypolicyspec-policy-v1beta1.json
endpointslicelist-discovery-v1beta1.json
localsubjectaccessreview-authorization-v1beta1.json
jobcondition-batch-v1.json
nonresourceattributes-authorization-v1beta1.json
csinode-storage-v1beta1.json
ingressclasslist-networking-v1beta1.json
ingressspec-networking-v1.json
endpoints-v1.json
jobtemplatespec-batch-v2alpha1.json
persistentvolumespec-v1.json
volumeattachmentstatus-storage-v1beta1.json
tokenrequest-authentication-v1.json
condition-meta-v1.json
prioritylevelconfigurationstatus-flowcontrol-v1alpha1.json
volumenoderesources-storage-v1.json
horizontalpodautoscalerlist-autoscaling-v2beta1.json
endpointport-discovery-v1beta1.json
clusterrolebindinglist-rbac-v1.json
nonresourcerule-authorization-v1.json
localobjectreference-v1.json
execaction-v1.json
glusterfsvolumesource-v1.json
handler-v1.json
statefulsetcondition-apps-v1.json
httpingressrulevalue-networking-v1.json
clusterrole-rbac-v1alpha1.json
nodestatus-v1.json
subjectaccessreviewstatus-authorization-v1beta1.json
ingresslist-networking-v1beta1.json
secretlist-v1.json
resourceattributes-authorization-v1.json
prioritylevelconfiguration-flowcontrol-v1alpha1.json
priorityclass-scheduling-v1beta1.json
ingressservicebackend-networking-v1.json
objectfieldselector-v1.json
customresourcesubresourcescale-apiextensions-v1beta1.json
storageclasslist-storage-v1.json
selinuxstrategyoptions-policy-v1beta1.json
apiservicecondition-apiregistration-v1.json
replicasetlist-apps-v1.json
ingressclasslist-networking-v1.json
webhookclientconfig-admissionregistration-v1beta1.json
prioritylevelconfigurationcondition-flowcontrol-v1alpha1.json
servicelist-v1.json
persistentvolumeclaimstatus-v1.json
metricspec-autoscaling-v2beta1.json
azurefilepersistentvolumesource-v1.json
servicereference-apiregistration-v1.json
csivolumesource-v1.json
persistentvolume-v1.json
leaselist-coordination-v1beta1.json
storageospersistentvolumesource-v1.json
intorstring-util-intstr.json
apigrouplist-meta-v1.json
certificatesigningrequestspec-certificates-v1beta1.json
selfsubjectrulesreviewspec-authorization-v1.json
servicespec-v1.json
cronjobspec-batch-v2alpha1.json
usersubject-flowcontrol-v1alpha1.json
ingressstatus-extensions-v1beta1.json
ingressspec-extensions-v1beta1.json
ingresstls-networking-v1beta1.json
ingressbackend-networking-v1beta1.json
apigroup-meta-v1.json
prioritylevelconfigurationlist-flowcontrol-v1beta1.json
storageosvolumesource-v1.json
persistentvolumestatus-v1.json
customresourcecolumndefinition-apiextensions-v1beta1.json
resourcemetricsource-autoscaling-v2beta1.json
iscsivolumesource-v1.json
queuingconfiguration-flowcontrol-v1beta1.json
labelselector-meta-v1.json
certificatesigningrequeststatus-certificates-v1.json
namespacelist-v1.json
jobspec-batch-v1.json
daemonsetstatus-apps-v1.json
limitrangeitem-v1.json
volumeattachmentspec-storage-v1alpha1.json
componentcondition-v1.json
ephemeralvolumesource-v1.json
prioritylevelconfigurationstatus-flowcontrol-v1beta1.json
customresourcedefinitioncondition-apiextensions-v1.json
containerresourcemetricsource-autoscaling-v2beta1.json
flowschemacondition-flowcontrol-v1beta1.json
limitresponse-flowcontrol-v1alpha1.json
daemonsetupdatestrategy-apps-v1.json
crossversionobjectreference-autoscaling-v2beta2.json
storageversioncondition-apiserverinternal-v1alpha1.json
subject-rbac-v1beta1.json
externalmetricstatus-autoscaling-v2beta1.json
subjectrulesreviewstatus-authorization-v1.json
tokenreview-authentication-v1beta1.json
namespacespec-v1.json
certificatesigningrequestcondition-certificates-v1beta1.json
servicereference-apiextensions-v1beta1.json
externaldocumentation-apiextensions-v1.json
all.json
configmapenvsource-v1.json
limitresponse-flowcontrol-v1beta1.json
selinuxoptions-v1.json
metricstatus-autoscaling-v2beta1.json
validatingwebhookconfiguration-admissionregistration-v1.json
horizontalpodautoscalercondition-autoscaling-v2beta2.json
groupsubject-flowcontrol-v1alpha1.json
objectmetricstatus-autoscaling-v2beta1.json
rolebindinglist-rbac-v1.json
horizontalpodautoscalerstatus-autoscaling-v2beta1.json
nodeselectorrequirement-v1.json
httpgetaction-v1.json
containerstatus-v1.json
resourcequotastatus-v1.json
policyruleswithsubjects-flowcontrol-v1alpha1.json
csinodespec-storage-v1beta1.json
cronjoblist-batch-v1beta1.json
keytopath-v1.json
podsecuritypolicy-policy-v1beta1.json
lifecycle-v1.json
volumeerror-storage-v1.json
loadbalancerstatus-v1.json
replicasetcondition-apps-v1.json
servicereference-admissionregistration-v1beta1.json
```
Can you see if the file exists for you and compare the contents, please?

jeffmaury commented:
I think it's a Windows issue as Paths.get(BASE_DIR, INDEX_FILE).toString() will not give something a class loader can load

adietish commented:
> Paths.get(
@jeffmaury 

> I think it's a Windows issue as Paths.get(BASE_DIR, INDEX_FILE).toString() will not give something a class loader can load

Looks as if you're right: https://stackoverflow.com/questions/33740388/paths-get-on-windows

adietish commented:
QE (@olkornii) did a new round of testing and reports that it's passing for him:
> didnt find nothing new

adietish commented:
@jeffmaury unit tests for `ResourceEditor` are done. Please review.

adietish commented:
@jeffmaury adding unit tests is finished. I consider this as ready to be merged unless you find bugs. Detailed testing instructions are here: https://github.com/redhat-developer/intellij-kubernetes/pull/178/files#diff-4434da0a1603ed9571e2ab9baca10c4fac6a50536889049f327deed186903207. Please test.

adietish commented:
@olkornii can you please confirm your approval by approving the github review so that it explicitly shows up?

olkornii commented:
@adietish sure, sorry.

jeffmaury commented:
I was not able to see the editor notification while document is modified in editor. Seems related to Windows path management as isResourceFile returns false:

![image](https://user-images.githubusercontent.com/695993/122964557-ae1b4300-d387-11eb-9464-18c3123f6e5d.png)
``

adietish commented:
intellij provides an editor that is similar in how it behaves: it allows to edit remote files: https://blog.jetbrains.com/phpstorm/2015/04/remote-edit-in-phpstorm-9-eap/. This editor is present in the base platform as `HttpFileEditor`
It's interesting to compare how they use notifications in the way I do. It can inspire us on how our editor should behave.

adietish commented:
replaced by #213 that includes the changes in this PR. Closing.
